### PR TITLE
refactor: Move FeatureSet to a separate package

### PR DESCRIPTION
**Pull request by michi-covalent**
Move FeatureSet to a separate package so that the sysdump package can import it to use the feature detection logic without introducing a circular dependency.

Ref: #1962

### CODE REVIEW

**Review by christarazi — APPROVED** [review]

**Review by aspsk — APPROVED** [review]

**Review by julianwiedmann — APPROVED** [review]

File: utils/features/features.go [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+	FeatureKNP Feature = "k8s-network-policy"
+
+	// Whether or not CIDR selectors can match node IPs
+	FeatureCIDRMatchNodes Feature = "cidr-match-nodes"
+
+	FeatureAuthSpiffe Feature = "mutual-auth-spiffe"
+
+	FeatureIngressController Feature = "ingress-controller"
+
+	FeatureEgressGateway Feature = "enable-ipv4-egress-gateway"
```
Comment (tklauser) [inline utils/features/features.go:60]: These names now stutter on usage, i.e. `features.FeatureSomething`. How about dropping the `Feature` prefix from these identifiers?

File: utils/features/features.go [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+// Feature is the name of a Cilium Feature (e.g. l7-proxy, cni chaining mode etc)
+type Feature string
+
+// FeatureStatus describes the status of a Feature. Some features are either
+// turned on or off (c.f. Enabled), while others additionally might include a
+// Mode string which provides more information about in what mode a
+// particular Feature is running ((e.g. when running with CNI chaining,
+// Enabled will be true, and the Mode string will additionally contain the name
+// of the chained CNI).
+type FeatureStatus struct {
```
Comment (tklauser) [inline utils/features/features.go:72]: Same remark about name stuttering here:

```suggestion
type Status struct {
```

File: utils/features/features.go [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+
+	if len(s.Mode) == 0 {
+		return str
+	}
+
+	return fmt.Sprintf("%s:%s", str, s.Mode)
+}
+
+// FeatureSet contains the FeatureStatus of a collection of Features.
+type FeatureSet map[Feature]FeatureStatus
```
Comment (tklauser) [inline utils/features/features.go:91]:
```suggestion
type Set map[Feature]Status
```

File: utils/features/features.go [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+			!fs[FeatureHostFirewall].Enabled) ||
+			fs[FeatureKPRHostPort].Enabled,
+	}
+
+	return nil
+}
+
+// FeatureRequirement defines a test requirement. A given FeatureSet may or
+// may not satisfy this requirement
+type FeatureRequirement struct {
```
Comment (tklauser) [inline utils/features/features.go:139]:
```suggestion
type Requirement struct {
```

File: utils/features/features.go [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+	requiresEnabled bool
+	enabled         bool
+
+	requiresMode bool
+	mode         string
+}
+
+// RequireFeatureEnabled constructs a FeatureRequirement which expects the
+// Feature to be enabled
+func RequireFeatureEnabled(feature Feature) FeatureRequirement {
```
Comment (tklauser) [inline utils/features/features.go:151]:
```suggestion
func RequireEnabled(feature Feature) Requirement {
```

File: utils/features/features.go [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+	return FeatureRequirement{
+		Feature:         feature,
+		requiresEnabled: true,
+		enabled:         true,
+	}
+}
+
+// RequireFeatureDisabled constructs a FeatureRequirement which expects the
+// Feature to be disabled
+func RequireFeatureDisabled(feature Feature) FeatureRequirement {
```
Comment (tklauser) [inline utils/features/features.go:161]:
```suggestion
func RequireDisabled(feature Feature) Requirement {
```

File: utils/features/features.go [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+	return FeatureRequirement{
+		Feature:         feature,
+		requiresEnabled: true,
+		enabled:         false,
+	}
+}
+
+// RequireFeatureMode constructs a FeatureRequirement which expects the Feature
+// to be in the given mode
+func RequireFeatureMode(feature Feature, mode string) FeatureRequirement {
```
Comment (tklauser) [inline utils/features/features.go:171]:
```suggestion
func RequireMode(feature Feature, mode string) Requirement {
```

File: utils/features/features.go [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+		Feature:      feature,
+		requiresMode: true,
+		mode:         mode,
+	}
+}
+
+// extractFeaturesFromConfigMap extracts features from the Cilium ConfigMap.
+// Note that there is no rule regarding if the default value is reflected
+// in the ConfigMap or not.
+func (fs FeatureSet) ExtractFeaturesFromConfigMap(ciliumVersion semver.Version, cm *v1.ConfigMap) {
```
Comment (tklauser) [inline utils/features/features.go:182]:
```suggestion
func (fs FeatureSet) ExtractFromConfigMap(ciliumVersion semver.Version, cm *v1.ConfigMap) {
```

File: utils/features/features.go [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+	fs[FeatureEgressGateway] = FeatureStatus{
+		Enabled: cm.Data["enable-ipv4-egress-gateway"] == "true",
+	}
+
+	fs[FeatureCIDRMatchNodes] = FeatureStatus{
+		Enabled: strings.Contains(cm.Data["policy-cidr-match-mode"], "nodes"),
+	}
+}
+
+func (fs FeatureSet) ExtractFeaturesFromNodes(nodesWithoutCilium map[string]struct{}) {
```
Comment (tklauser) [inline utils/features/features.go:251]:
```suggestion
func (fs FeatureSet) ExtractsFromNodes(nodesWithoutCilium map[string]struct{}) {
```

**Review by tklauser — APPROVED** [review]

**Pull request merged**